### PR TITLE
feat(MV-383): Port SGDContiguous

### PR DESCRIPTION
**Pull request by HoangNguyenMoreh**
Ticket link: https://moreh.atlassian.net/browse/MV-383

This PR ports the `SGDcontiguous` OpenCL kernel to MIOpen. 

Unit tests:
image

Driver test:
image

<details>
<summary>float16</summary>
<p>

| dtype   | dims             | lr     | momentum | dampening | weight_decay | direction | ROCm   | MIOpen HIP | ROCm/MIOpen  |
|---------|------------------|--------|----------|-----------|--------------|-----------|--------|------------|--------------|
| float16 |32 3 0 3 3   |0.004 |0.9 |0       |0      |fwd|19508|5316 |3.48|
| float16 |32 0 0 0 0   |0.004 |0.9 |0       |0      |fwd|19988|5743 |3.66|
| float16 |32 32 3 3 3  |0.004 |0.9 |0       |0      |fwd|42472|6436 |6.59|
| float16 |64 32 3 3 3  |0.004 |0.9 |0       |0      |fwd|62332|6490 |9.60|
| float16 |32 0 0 0 0   |0.004 |0.9 |0       |0      |fwd|17635|5494 |3.20| 
| float16 |32 3 0 3 3   |0.004 |0.9 |0       |0      |fwd|19060|2596 |7.34| 
| float16 |32 32 3 3 3  |0.004 |0.9 |0       |0      |fwd|41800|2169 |19.27| 
| float16 |64 32 3 3 3  |0.004 |0.9 |0       |0      |fwd|62204|7076 |8.79| 
| float16 |61 3 0 11 11 |0.01  |0.9 |0       |0.0005 |fwd|85905|2151 |39.93|
| float16 |192 64 0 5 5 |0.01  |0.9 |0       |0.0005 |fwd|75823|8908 |8.51|
| float16 |61 3 0 11 11 |0.01  |0.9 |0       |0.0005 |fwd|85297|6845 |12.46| 
| float16 |192 64 0 5 5 |0.01  |0.9 |0       |0.0005 |fwd|75583|9868 |7.65| 
| float16 |64 3 0 3 3   |0.01  |0.9 |0.0005  |0      |fwd|15331|2596 |5.90|
| float16 |64 64 0 1 1  |0.01  |0.9 |0.0005  |0      |fwd|17700|2293 |7.71|
| float16 |64 3 0 3 3   |0.01  |0.9 |0.0005  |0      |fwd|15299|2524 |6.06| 
| float16 |64 64 0 1 1  |0.01  |0.9 |0.0005  |0      |fwd|17604|6756 |2.60| 

</p>
</details>

<details>
<summary>float32</summary>
<p>

| dtype   | dims             | lr     | momentum | dampening | weight_decay | direction | ROCm   | MIOpen HIP | ROCm/MIOpen  |
|---------|------------------|--------|----------|-----------|--------------|-----------|--------|------------|--------------|
| float32 |32 0 0 0 0   |0.004 |0.9 |0       |0      |fwd|19236|2969|6.47|
| float32 |32 3 0 3 3   |0.004 |0.9 |0       |0      |fwd|20084|5636|3.56|
| float32 |32 32 3 3 3  |0.004 |0.9 |0       |0      |fwd|45641|6738|6.77|
| float32 |64 32 3 3 3  |0.004 |0.9 |0       |0      |fwd|66397|6810|9.74|
| float32 |32 0 0 0 0   |0.004 |0.9 |0       |0      |fwd|17955|3058|5.87| 
| float32 |32 3 0 3 3   |0.004 |0.9 |0       |0      |fwd|19156|6650|2.88| 
| float32 |32 32 3 3 3  |0.004 |0.9 |0       |0      |fwd|43480|2062|21.08| 
| float32 |64 32 3 3 3  |0.004 |0.9 |0       |0      |fwd|65885|7556|8.71| 
| float32 |61 3 0 11 11 |0.01  |0.9 |0       |0.0005 |fwd|86865|2080|41.76|
| float32 |192 64 0 5 5 |0.01  |0.9 |0       |0.0005 |fwd|80848|9761|8.28|
| float32 |61 3 0 11 11 |0.01  |0.9 |0       |0.0005 |fwd|83056|2134|38.92| 
| float32 |192 64 0 5 5 |0.01  |0.9 |0       |0.0005 |fwd|80992|2720|29.77| 
| float32 |64 3 0 3 3   |0.01  |0.9 |0.0005  |0      |fwd|15459|2756|5.60|
| float32 |64 64 0 1 1  |0.01  |0.9 |0.0005  |0      |fwd|18163|6401|2.83|
| float32 |64 3 0 3 3   |0.01  |0.9 |0.0005  |0      |fwd|15619|6721|2.32| 
| float32 |64 64 0 1 1  |0.01  |0.9 |0.0005  |0      |fwd|17715|2293|7.72| 

</p>
</details>

<details>
<summary>bfloat16</summary>
<p>

| dtype   | dims             | lr     | momentum | dampening | weight_decay | direction | ROCm   | MIOpen HIP | ROCm/MIOpen  |
|---------|------------------|--------|----------|-----------|--------------|-----------|--------|------------|--------------|
| bfloat16 |32 0 0 0 0   |0.004 |0.9 |0       |0      |fwd|20180|5760|3.50|
| bfloat16 |32 3 0 3 3   |0.004 |0.9 |0       |0      |fwd|19972|5458|3.65|
| bfloat16 |32 32 3 3 3  |0.004 |0.9 |0       |0      |fwd|46793|2080|22.49|
| bfloat16 |64 32 3 3 3  |0.004 |0.9 |0       |0      |fwd|68942|2098|32.86|
| bfloat16 |32 0 0 0 0   |0.004 |0.9 |0       |0      |fwd|17780|6223|2.85| 
| bfloat16 |32 3 0 3 3   |0.004 |0.9 |0       |0      |fwd|19604|2631|7.45| 
| bfloat16 |32 32 3 3 3  |0.004 |0.9 |0       |0      |fwd|44361|6934|6.39| 
| bfloat16 |64 32 3 3 3  |0.004 |0.9 |0       |0      |fwd|68622|6863|9.99| 
| bfloat16 |61 3 0 11 11 |0.01  |0.9 |0       |0.0005 |fwd|113542|2133|53.23|
| bfloat16 |192 64 0 5 5 |0.01  |0.9 |0       |0.0005 |fwd|84960|9014|9.42|
| bfloat16 |61 3 0 11 11 |0.01  |0.9 |0       |0.0005 |fwd|112950|6969|16.20| 
| bfloat16 |192 64 0 5 5 |0.01  |0.9 |0       |0.0005 |fwd|85697|2880|29.75| 
| bfloat16 |64 3 0 3 3   |0.01  |0.9 |0.0005  |0      |fwd|15571|5476|2.84|
| bfloat16 |64 64 0 1 1  |0.01  |0.9 |0.0005  |0      |fwd|17923|6383|2.80|
| bfloat16 |64 3 0 3 3   |0.01  |0.9 |0.0005  |0      |fwd|15651|6276|2.49| 
| bfloat16 |64 64 0 1 1  |0.01  |0.9 |0.0005  |0      |fwd|19028|2311|8.23| 

</p>
</details>

Geomean of all cases
| type| direction | geomean |
|-----------|-------------|--------|
| float16 |fwd|6.93 |
| float32 |fwd|7.33 |
| bfloat16 |fwd|7.27 |

Checklist:
- [x] Kernel ported from OpenCL
- [x] MIOpen frontend code (prepare/launch the kernel)
- [x] CPU implementation of the operation (for unit tests)
- [x] gtest unit tests (working & passing verification)
- [x] MIOpenDriver driver code (working & passing verification `-V`)
- [x] Code cleanup

### CODE REVIEW

**Comment by kyeonghwanryu** [comment]
I see the "Code Cleanup" is still unchecked.
I suggest you to keep this PR as a draft before it's done.

**Comment by HoangNguyenMoreh** [comment]
> I see the "Code Cleanup" is still unchecked. I suggest you to keep this PR as a draft before it's done.

@kyeonghwanryu Thank you for your suggestion. Actually, it's not [Code cleanup] problem, I'm facing an issue about unit tests, the test-run not always passed.

**Comment by HoangNguyenMoreh** [comment]
@kyeonghwanryu It's turn out that `miopen-1` node is having some problem with the devices 
image

I've re-implemented on `miopen-2` and successfully ran the tests.

File: driver/CMakeLists.txt [hunk 1]
```diff
@@ -50,6 +50,7 @@ add_executable(MIOpenDriver
     dm_rnn.cpp
     dm_softmax.cpp
     dm_sum.cpp
+    dm_SGD.cpp
```
Comment (kyeonghwanryu) [inline driver/CMakeLists.txt]: Please follow the capitalization for the names in other cases.
For example in this case, dm_sgd.cpp would be right.

File: src/SGD.cpp [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+                          const TensorDescriptor& momentumBufferOutDesc,
+                          Data_t momentumBufferOut,
+                          double lr,
+                          double momentum,
+                          double dampening,
+                          double weightDecay,
+                          char nesterov,
+                          char momentum_initialized)
+{
+    const auto problem = SGD::ProblemDescription{paramInDesc, paramOutDesc, gradDesc, momentumBufferInDesc, momentumBufferOutDesc, lr, momentum, dampening, weightDecay, nesterov, momentum_initialized};
```
Comment (et16kr) [inline src/SGD.cpp]: Please set clang-format-12
```
clang-format-12 -i src/SGD.cpp
```

File: driver/CMakeLists.txt (hunk 1 above)
Comment (HoangNguyenMoreh) [reply driver/CMakeLists.txt]: In the modnn, the opencl file name is `SGD.cl` so I thought we will capitalize all the `SGD` characters in files and code.
I will update all the file name from `...SGD...` to `...sgd...`. (except for `MIOpenSGD.cpp`)
Thank you!

File: src/SGD.cpp (hunk 2 above)
Comment (HoangNguyenMoreh) [reply src/SGD.cpp]: My mistake, thank you!

File: src/kernels/MIOpenSGD.cpp [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+ *
+ *******************************************************************************/
+#ifndef MIOPEN_DONT_USE_HIP_RUNTIME_HEADERS
+#include <hip/hip_fp16.h>
+#include <hip/hip_runtime.h>
+#endif
+
+#include "float_types.h"
+
+#if MIOPEN_USE_BFP16 == 1
```
Comment (kyeonghwanryu) [inline src/kernels/MIOpenSGD.cpp]: Why do you need this?
They are already defined at float_types.h.

**Comment by kyeonghwanryu** [comment]
Please add geomean improvements by type to the PR's description.

File: src/kernels/MIOpenSGD.cpp (hunk 3 above)
Comment (HoangNguyenMoreh) [reply src/kernels/MIOpenSGD.cpp]: Thanks, I see.
I will remove this part.

**Comment by kyeonghwanryu** [comment]
Good work. You can change the ticket to done.
Perf Improvement looks very good.

**Pull request closed without merging**